scipy/stats/stats.py

```diff
@@ -2465,15 +2465,20 @@ def gstd(a, axis=0, ddof=1):
             raise ValueError(
                 'Infinite value encountered. The geometric standard deviation '
                 'is defined for strictly positive values only.')
-        elif np.less_equal(a, 0).any():
+        a_nan = np.isnan(a)
+        a_nan_any = a_nan.any()
+        if a_nan.any():
+            return np.exp(np.std(log(a, where=~a_nan), axis=axis, ddof=ddof))
+        elif ((a_nan_any and np.less_equal(a[~a_nan], 0).any()) or
```

Comment:
With the above if statement ``a_nan_any`` will always be False. So the statement becomes:

``(False and np.less_equal(a[~a_nan], 0).any()) or (not False and np.less_equal(a, 0).any()``
<==> ``(False) or (np.less_equal(a, 0).any())``
<==> ``np.less_equal(a, 0).any()``.

Without the above if statement it looks like a good solution.

Reply:
I am not sure I follow. The conditional reads:

```
( (a_nan_any and np.less_equal(a[~a_nan], 0).any() ) or 
   (not a_nan_any and np.less_equal(a, 0).any() )   )
```
If `a_nan_any` is True, we apply mask selector and only check non-nan numbers with zero. 

if `a_nan_any` is False, we avoid applying the mask, and associated extra copying, and compare all elements with zero.

I thought, however, that I had removed the previous `if ` statement, but evidently not.

Reply:
I pushed the change deleting the leading `if` branch.

Reply:
I could have made that clearer.

The code goes something like this:

```python
a_nan = np.isnan(a)
a_nan_any = a_nan.any()
if a_nan.any():
    # do stuff....
# If we get to here  a_nan.any() is False.
# As a_nan_any = a_nan.any(), a_nan_any will also be False.
elif ((a_nan_any and np.less_equal(a[~a_nan], 0).any()) or
       (not a_nan_any and np.less_equal(a, 0).any())):
    # handle negativity check
```
If ``a_nan.any()``  is True the first block is run, so, to reach the second statement, ``a_nan.any()`` must be False. As ``a_nan_any = a_nan.any()``, ``a_nan_any`` must also be False, otherwise the second would not be reached.

Reply:
Yes, thank you. I realized that moments after I posted the comment, and pushes changes removing the 'if a_nan_any:` branch of the if statement.